netty/src/main/java/io/grpc/netty/NettyClientHandler.java

```diff
@@ -380,6 +398,28 @@ private void onHeadersRead(int streamId, Http2Headers headers, boolean endStream
     if (streamId != Http2CodecUtil.HTTP_UPGRADE_STREAM_ID) {
       NettyClientStream.TransportState stream = clientStream(requireHttp2Stream(streamId));
       PerfMark.event("NettyClientHandler.onHeadersRead", stream.tag());
+      // check metadata size vs soft limit
+      int h2HeadersSize = Utils.getH2HeadersSize(headers);
+      boolean shouldFail =
+          Utils.shouldRejectOnMetadataSizeSoftLimitExceeded(
+              h2HeadersSize, softLimitHeaderListSize, maxHeaderListSize);
+      if (shouldFail && endStream) {
+        throw Status.RESOURCE_EXHAUSTED
```

Comment:
I don't think this works like you think it works, and there's no test to confirm it is working as intended.  I think this will end up calling onConnectionError() and disconnecting. I think you need to call `stream.transportReportStatus()` and send RST_STREAM if the stream hasn't been closed. Deserving of a test.

Reply:
Done.